.github/workflows/haskell.yml

```
@@ -630,3 +630,18 @@ jobs:
           echo "Commit $TAG was not found on the master-artifacts branch of formal-ledger-specifications."
           exit 1
         }
+  undefined-check:
+    name: Check that there are no `undefined`s in the code
+    runs-on: ubuntu-latest
+    if: ${{ github.base_ref != '' && github.ref != '' }}  # Only true for PRs
+
+    steps:
+    - uses: actions/checkout@v4
+    - name: Check for `undefined`s in the diffs
+      run: |
+        PR_TARGET=${{ github.base_ref }}
+        git fetch origin -n --refmap= +$PR_TARGET:pr-target
+        if git diff pr-target "*.hs" | grep '^\+.*undefined'; then
```

A minor issue, and probably not worth changing, but maybe useful for future reference:

```suggestion
        if git diff pr-target '*.hs' | grep '^\+.*undefined'; then
```

In general, use single-quotes unless you need to interpolate variable values into the string. It's easy to put a literal `$` into the string, now or later, and forget to change the quotes.

Also, note that this will pick up the word `undefined` if it was added in a comment. However, I don't think there's an easy way to make it more precise.

---

I found this [SO Question](https://stackoverflow.com/questions/52063243/can-haskell-compiler-give-warning-on-functions-that-used-undefined) about ways to make it more precise, and it seems the two options are to use `hlint` with a custom rule, and to use a custom prelude. I know @lehins has expressed a liking for the latter, for other reasons.

---

I found a way to do it with `hlint` that ignores all other problems:

```shell
cat <<EOF >.hlint-undefined
- ignore: {}
- error:
    name: "Don't use undefined"
    lhs: "undefined"
    rhs: 'error "We have a problem"'
EOF
git diff --name-only pr-target | xargs -d '\n' hlint -h .hlint-undefined.yaml
```

You then don't need the `if` block because `hlint` will output the necessary message.

We could of course check the `.hlint-undefined` file into the repo instead of creating it on the fly.

Using this, I found that we have a lot of uses of `undefined` in the codebase, almost all of them in `eras/byron`. The three others were in:

* `libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Coders.hs`
* `libs/vector-map/test/Test/VMap.hs`

Of course, this won't matter unless there's a change to one of those files. However, `hlint` looks at the whole file rather than the changes, so it would be necessary to fix the uses of `undefined` in the unchanged parts, whereas the `grep` approach doesn't require that.

---

We don't need to fix all the current uses of `undefined`, so I am more in favor of the grep approach

---

I could switch from grep to [tree sitter](https://tree-sitter.github.io/tree-sitter/). It can tell whether something is a comment or a term.